fix: Check elevation values before replacing parent : LayeredMaterialNodeProcessing

nosy-b opened the pull request:
This small fix allows for better handling of multiple DTM, it answers #192.

- WMTS can send some tiles with no real value in it through a code like -99999. We check this before replacing the elevation parent tile so we don't create holes and can keep other less precise DTM tile, typically around their borders
- This PR also modifies the index to handle a world DTM and a more precise for France.

It is not elegant at all, it should be handled with LayerState for example with a state specifying when an elevation tile is its max level.

The same should be done for Imagery as sometimes WMTS send white images, not errors.

peppsac commented on index.html:
```diff
@@ -47,8 +47,8 @@
             itowns.Fetcher.json('examples/layers/JSONLayers/Region.json').then(result => globeView.addLayer(result));
-
-            itowns.Fetcher.json('examples/layers/JSONLayers/IGN_MNT.json').then(result => globeView.addLayer(result));
+        
```
Whitespace change

peppsac commented on src/Core/Layer/LayerUpdateState.js:
```diff
@@ -3,6 +3,7 @@ const UPDATE_STATE = {
     PENDING: 1,
     ERROR: 2,
     DEFINITIVE_ERROR: 3,
+    DEEPEST_LEVEL: 4,
```
This doesn't seem to be used anywhere?

nosy-b replied:
Yes it comes from a test, I should remove it

peppsac commented on src/Core/Scheduler/Providers/IoDriver_XBIL.js:
```diff
@@ -49,10 +49,10 @@ IoDriver_XBIL.prototype.computeMinMaxElevation = function computeMinMaxElevation
     if (max === -1000000 || min === 1000000) {
         return { min: undefined, max: undefined };
     }
+
```
Unneeded whitespace changes

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -236,11 +236,14 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
         currentElevation = material.getElevationLayerLevel();
     }
+
+
```
Whitespaces...

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -281,6 +284,8 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
 
             node.layerUpdateState[layer.id].success();
+            // eslint-disable-next-line no-console
+            // if (terrain.texture.image.data[0] < 0) console.log('AAAAAAAAAAAAAAAAAAAFCefzfzefzefezfezfzefzef');
```
Maybe you should remove this :)

autra replied:
Oh nice one. Taking this to replace my `coucou` or `wtf` ;-)

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -289,7 +294,7 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
                 terrain.texture.needsUpdate = true;
             }
-            node.setTextureElevation(terrain);
+            if (terrain.texture.image.data[0] && terrain.texture.image.data[terrain.texture.image.data.length - 1] > -99999) node.setTextureElevation(terrain);
```
Hmm I'm not really a fan of this proposal. I've seen some place where data[0] was valid but data[15] wasn't for instance.

Also have you considered this alternative: allow several textures layers?
They wouldn't blend their values like colors layers but instead the first elevation texture with a valid value would be used.
So in your case, you would declare `IGN_MNT_HIGHRES` as the first elevation layer, and `MNT_WORLD_SRTM3` as the second one and it would act as a fallback when `HIGH_RES` gives no value.

WDYT?

autra replied:
(nits, formatting: this line is really long, prefer the
```js
if (...) {
    ....
}
``` 
form)

nosy-b replied:
The test is not complete, I wanted to test diagonal extremity pixels. More like:
```
let terrainData = terrain.texture.image.data;
if (terrainData[0] > -99999 && terrainData[terrainData.length - 1] > -99999) 
        node.setTextureElevation(terrain);
```

It's a quick fix to avoid using non-real elevation tiles value that create holes.
As you asked @autra the IGN_MNT_HIGHRES DTM is complementary with the WORLD_DTM now that replaces the low res IGN_MNT (We should rename all MNT with DTM also).
So when declaring them, their TileMatrixSet have an empty union. I should try to see how it works when we don't take care of that.

@peppsac Yes I think it should be done as we do for color layers, with an order sequence for elevation layers. To avoid overloading the number of textures, the active elevation layer choice should be done before sending elevation to the shader.

nosy-b replied:
_+_ This fix goes with a rapid V2 release and can be kept in the future. 
_+_ Very few users will mix many DTM on the same region
 _-_   It will certainly need a more strong and generic approach as we mentioned above (2.1?)

peppsac replied:
Can you update the PR with the whitespaces fixes plus:
- a test of the 4 corners of the terrainData
- a comment stating that this is a compromise and that we should implement multi elevation layer instead
- make sure this works with texture images (see planar example)

With these changes this PR should be mergeable.

peppsac reviewed (commented):
Thanks for taking care of this issue.
I've left a few minor style comment and a general design one :)

autra commented on examples/layers/JSONLayers/IGN_MNT_HIGHRES.json:
```diff
@@ -7,7 +7,7 @@
 		"type": 1,
 		"options": {
-			"groups": [3, 7, 11, 14]
+			"groups": [11, 14]
```
Just  a little question: why `IGN_MNT_HIGHRES.json` is modified at all? Let's either keep it as it is if we plan on use it, or let's remove it entirely.

gchoqueux commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -92,6 +92,16 @@ function refinementCommandCancellationFn(cmd) {
+    // We check if the elevation texture has some significant values through corners
+    const tData = texture.image.data;
+    const l = tData.length;
+    return tData[0] > -99999 &&
```
use `noDataValue` in layer option

gchoqueux commented on examples/layers/JSONLayers/WORLD_DTM.json:
```diff
@@ -0,0 +1,69 @@
+    "updateStrategy": {
+        "type": 1,
+        "options": {
+            "groups": [3, 7, 10]
```
may be use 
for WORLD
`"groups": [3, 7]`
and for HIGH RES
`"groups": [10, 14]`

nosy-b committed:
fix: Check elevation values before replacing parent : LayeredMaterialNodeProcessing

Quick check to avoid using elevation texture with no significant values
We should later implement multi elevation layer to choose the one to use at each level

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -281,15 +291,17 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
             }
 
             node.layerUpdateState[layer.id].success();
-
```
Nit: unneeded change

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -281,15 +291,17 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
                 terrain.texture.needsUpdate = true;
             }
-            node.setTextureElevation(terrain);
+            if (!terrain.texture.image.data || checkNodeElevationTextureValidity(terrain.texture, layer.noDataValue))
```
Maybe `checkXbilTextureValidity` would be more appropriate?

And what about calling: 
```js
node.layerUpdateState[layer.id].failure(Date.now(), true);
```
to indicate that this `layer` will never supply a valid texture for this node?

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -281,15 +291,17 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, force)
+            if (!terrain.texture.image.data || checkNodeElevationTextureValidity(terrain.texture, layer.noDataValue))
+                // Quick check to avoid using elevation texture with no significant values
+                // We should later implement multi elevation layer to choose the one to use at each level
+                { node.setTextureElevation(terrain); }
```
Style issue:
```js
if (...) {
  // Quick check ...
  // ...
  node.setTextureElevation(terrain);
}
```

peppsac commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -92,6 +92,16 @@ function refinementCommandCancellationFn(cmd) {
+    const l = tData.length;
+    return tData[0] > noDataValue &&
+           tData[l - 1] > noDataValue &&
+           tData[Math.sqrt(l) - 1] > noDataValue &&
```
I'd prefer `tData[width - 1]` and `tData[height * width]` instead of `Math.sqrt` usage but it's up to you

nosy-b commented:
Ok, I changed a bit to handle more cleanly those no data value containing tiles. 
They now use parent values for the pixel with no-data-values when we detect that it's not a normal tile so it doesn"t slow down the rendering.
I 'll rebase

peppsac reviewed (approved):
Seems good.
Can you rebase this PR please?

nosy-b commented:
@peppsac rebased, thx.
